### PR TITLE
[Blazor] Support for primitive types in [SupplyParameterFromForm]

### PR DESCRIPTION
**Pull request by javiercn**
* Adds support for binding IParsable<T> values and Nullable values.

FormDataMapper is the entry point for mapping form values into a type, it receives a FormDataReader and FormDataMapperOptions. The options contain an internal collection of converters and factories that are used to define how to bind a given type.

For well-known types we prepopulate the list of converters.

* FormDataConverter <-> ModelBinder.
* FormDataConverterFactory <-> ModelBinderProvider

https://github.com/dotnet/aspnetcore/issues/46688#issuecomment-1570315950

Some basic benchmark data against MVC. Beware that this is with any of the main optimizations.

|                                Method |     Mean |    Error |   StdDev |     Op/s | Ratio | RatioSD | Gen 0 | Gen 1 | Gen 2 | Allocated |
|-------------------------------------- |---------:|---------:|---------:|---------:|------:|--------:|------:|------:|------:|----------:|
|       ModelBinding_PrimitiveTypes_Mvc | 43.08 μs | 1.120 μs | 3.141 μs | 23,212.5 |  1.00 |    0.00 |     - |     - |     - |      3 KB |
| ModelBinding_PrimitiveType_Components | 25.11 μs | 0.834 μs | 2.367 μs | 39,821.0 |  0.59 |    0.07 |     - |     - |     - |      2 KB |

### CODE REVIEW

File: src/Components/Endpoints/src/Binding/FormDataConverterOfT.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace Microsoft.AspNetCore.Components.Endpoints.Binding;
+
+internal abstract class FormDataConverter<T> : FormDataConverter
+{
+    internal abstract bool TryRead(ref FormDataReader context, Type type, FormDataMapperOptions options, out T? result, out bool found);
```
Comment (MackinnonBuck) [inline src/Components/Endpoints/src/Binding/FormDataConverterOfT.cs:8]: Is `type` ever going to be different from `typeof(T)`?

**Review by MackinnonBuck — APPROVED** [review]
Just a couple questions but this looks great to me.

File: src/Components/Endpoints/src/Binding/FormDataMapper.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    {
+        try
+        {
+            var converter = options.ResolveConverter<T>();
+            if (converter.TryRead(ref reader, typeof(T), options, out var result, out _))
+            {
+                return result;
+            }
+
+            // We don't do error handling yet.
```
[the remark this answers is not in the document]
Comment (javiercn) [reply src/Components/Endpoints/src/Binding/FormDataMapper.cs:20]: It's covered as part of https://github.com/dotnet/aspnetcore/issues/46688

File: src/Components/Endpoints/src/Binding/FormDataConverterOfT.cs (hunk 1 above)
Comment (javiercn) [reply src/Components/Endpoints/src/Binding/FormDataConverterOfT.cs:8]: It's not obvious yet in this PR, but it's passed in for a different reason. When you do `typeof(T)` the generated code needs to ask the type system at runtime (this is in IL) for the type instance, which involves a lookup. For that reason, many serializers cache and pass the type instance along.

If you look at System.Text.Json, they do something similar. See [here](https://learn.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonconverter-1.read?view=net-7.0#system-text-json-serialization-jsonconverter-1-read(system-text-json-utf8jsonreader@-system-type-system-text-json-jsonserializeroptions))

The other part is that we might still need this if in the future we want to support polymorphic binding or enable people to do so, so we can't just assume typeof(T) == type in general.

**Pull request merged**